enhancements/builds/volume-secrets.md

```diff
@@ -0,0 +1,129 @@
+
+For dockerfile builds, the user is instructed to add appropriate `ADD` and `RUN rm` commands to their dockerfile to inject the content that is available in the build's working directory (along with their application source, where applicable).
+
+There are a few undesirable aspects to this:
+1) In the dockerfile case, the content can still be found in lower layers of the image unless a layer squashing option is 
+selected.
+2) Requires extra work by the user in the Dockerfile, so each Dockerfile must be customized
+
+This enhancement proposes to use buildah's capability to mount a volume at build time.  The content mounted into the build pod would be then mounted into the container processing the Dockerfile, making that content available within the container so Dockerfile commands could reference it.  No explicit `ADD` would be required, and since mounted content is not committed to the resulting image, no `RUN rm` equivalent is required to clean up the injected content.
+
+
+## Motivation
+
+### Goals
+
+* Simplify how users consume secret + configmap content in builds
+* Increase the security of protected content being injected to images
+* Simplify use cases like mounting subscription entitlement credentials into a build so subscription content can be pulled down during a build while not including the credentials in the resulting image.
+
+### Non-Goals
+
+* This enhancement is not related to the frequently requested "build volumes" feature which would make it possible to mount a volume during a build and read/write content on that volume during the build such as to cache dependencies.
+
+
+## Proposal
+
+### User Stories [optional]
+
+The enabled use cases are essentially identical to what can be done with the configmap/secret input api in builds today, but with a better user experience and security as discussed above.  It does not enable a new use case that is not already possible today, except that layer squashing will not be required.
+
+### Implementation Details/Notes/Constraints [optional]
+
+We will need to introduce a new field in the build api which allows the user to indicate that they want to use this mechanism to inject their configmap/secret content, rather than the existing mechanism.
+
+The build pod will continue to mount the content as it already does, however the logic that generates the dockerfile(for s2i) and invokes buildah(for both dockerfile + s2i builds) will need to change to not augment the dockerfile, but instead to pass volume mount paths to buildah.  It will mount the content from the path it was mounted into the build pod, and mount it to the destination path provided by the user (the same path where the logic previously injected it via the dockerfile `ADD` command).
+
+proposed api:
+
+```
+apiVersion: v1
+items:
+- apiVersion: build.openshift.io/v1
+  kind: BuildConfig
+  metadata:
+    name: mybuild
+    namespace: p1
+  spec:
+    source:
+      secrets:
```

Comment:
I don’t like this.  I would want to mount volumes in the future design.  Mounting volumes and injecting the current set of secrets are already different (in terms of use case)

Reply:
hm.  So you're looking for an api which is "volume as an input" and then that volume can potentially come from a configmap, secret, PV, etc.

And then after we mounted that content into the build-pod, we'd wire it through as a buildah-mount.

It can be done, i think it's substantially more work than what this augmentation would require (define+implement an entire new api for mounting content into the build-pod, which this enhancement doesn't need since all that logic already exists for secret/configmap input api).

and of course we'll have to constrain the initial implementation to only allow secret+configmap "volume-types".

but i will take a pass at it

Reply:
I think volume mounts aren't part of "source" - they are part of "build execution environment" (which admittedly we don't have well defined).  I think source is about assembly of the input, but execution environment is more about the pod.

Reply:
reasonable.  unfortunately today the closest thing we have to something that's part of the execution environment is our "env var" injection, which is defined on a per-strategy basis.

i'd prefer not to do this on a per-strategy basis, though we certainly could.

if we did it top-level it probably looks like:

```
apiVersion: v1
items:
- apiVersion: build.openshift.io/v1
  kind: BuildConfig
  metadata:
    name: mybuild
    namespace: p1
  spec:
    source:
      secrets:
      - secret: 
          name: myOtherSecret
        destinationDir: /tmp/creds
      - configMaps:
        configMap:
          name: myOtherConfigMap
        destinationDir: /tmp/config
    executionContext:
      volumes:
      - type: Secret
        reference:
          name: mySecretName
        destinationDir: /etc/subscriptions
      - type: ConfigMap
        reference: 
          name: myConfigMapName
        destinationDir: /tmp/config
```


if we bolted it into strategy we'd have:
```
apiVersion: v1
items:
- apiVersion: build.openshift.io/v1
  kind: BuildConfig
  metadata:
    name: mybuild
    namespace: p1
  spec:
    source:
      git:
        uri: https://github.com/openshift/nodejs-ex
      secrets:
      - secret: 
          name: myOtherSecret
        destinationDir: /tmp/creds
      - configMaps:
        configMap:
          name: myOtherConfigMap
        destinationDir: /tmp/config
    strategy:
      type: Source
      sourceStrategy:
        env:
        - name: foo
          value: bar
        volumes:
        - type: Secret
          reference:
            name: mySecretName
          destinationDir: /etc/subscriptions
        - type: ConfigMap
          reference: 
            name: myConfigMapName
          destinationDir: /tmp/config
```